### PR TITLE
Installer pip coexistence via PATH precedence + CI-only release orchestrator

### PR DESCRIPTION
**Pull request by sammy-vastai**
Two pieces of installer/release cleanup:

## fix(installer): resolve pip coexistence via PATH precedence (env.sh pattern)

The rc file gets one constant line — `[ -f "$HOME/.vastai/env.sh" ] && . "$HOME/.vastai/env.sh"` — appended once, never rewritten (the rustup/uv pattern). `~/.vastai/env.sh` is regenerated on every install and carries the logic: prepend `~/.local/bin` unless it's already first, then source shell-appropriate tab completion. Precedence over a pip-installed `vastai` (the pip package ships one for the SDK) is therefore asserted at every shell startup — a pip install appearing later is out-ranked with no re-run. The coexistence warning survives only where the rc line couldn't be written: `--no-modify-path`, non-interactive runs, or an unrecognized shell.

Because rc files are append-only with a constant line, there is nothing to rewrite and no marker blocks to corrupt — the earlier marker-rewrite machinery (and its symlink, malformed-block, and idempotency edge cases flagged in review) is deleted rather than hardened.

Tests: `env_sh` (content), `pip_shadowing` (functional: sourcing env.sh out-ranks a pip fake), `pip_shadowing_quiet` (pty install resolves silently), `pip_shadowing_rerun` (idempotent re-run through a symlinked rc, no spurious warning), `rc_safety` (never edits rc non-interactively; exact-line idempotency). 14 tests pass; shellcheck -S style clean.

## chore(release): make publish_release.py CI-only, drop local mode

The local release mode (tag/push, PyPI polling, interactive confirms, post-publish verify) duplicated what CI owns end-to-end, so it's removed (262 → 138 lines). The script does exactly what `python-publish.yml` runs on tag push: hash the `dist/` wheel → render manifest → draft Release → upload assets → flip live. `--dry-run` prints the exact argv of every command. The manual runbook in `docs/install-design.md` §13 remains as break-glass.

### CODE REVIEW

**Comment by sammy-vastai** [comment]
Both Copilot findings addressed in 4b0fbf9:

- **rc marker idempotency**: the marker check is now content-aware — when the existing block predates a newly needed PATH export (written while PATH was healthy, foreign `vastai` appeared later), a re-run rewrites the block between markers to reassert precedence instead of returning early. Covered by new `test_pip_shadowing_rerun` (pty re-run over a markered rc: block gains the export, no duplicate markers, completion survives, no warning). If the end marker is missing (corrupt block) the installer leaves the rc alone and the coexistence warning covers it.
- **`gh` existence probe**: now raises a clean `ReleaseError("required tool not found: gh")` instead of a raw `FileNotFoundError` traceback, consistent with every other call site.

All 14 installer tests pass; shellcheck -S style clean.

File: scripts/publish_release.py [hunk 1]
```diff
@@ -177,8 +88,11 @@ def attach_release(tag, manifest_dir, wheel, *, dry):
...
-    exists = subprocess.run(["gh", "release", "view", tag],
-                            capture_output=True, text=True).returncode == 0
+    try:
+        exists = subprocess.run(["gh", "release", "view", tag],
+                                capture_output=True, text=True).returncode == 0
+    except FileNotFoundError:
+        raise ReleaseError("required tool not found: gh")
     if not exists:
         run(["gh", "release", "create", tag, "--draft", "--title", tag,
              "--notes", f"{PACKAGE} {tag.lstrip('v')}", "--generate-notes"], dry=dry)
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/publish_release.py:99]: Declined — the flags compose by design: `gh release create --help` documents that notes passed via `--notes` are **prepended** to the auto-generated ones. Empirically, this exact command (unchanged by this PR) created v1.2.1 this morning; its release body is `vastai 1.2.1` followed by the generated "What's Changed" section. Dropping `--notes` as suggested would remove the `vastai X.Y.Z` header from every future release body.

File: scripts/install.sh [hunk 2]
```diff
@@ -270,32 +271,57 @@ setup_path() {
...
+    # Marker idempotency is content-aware: an existing block written when PATH
+    # was healthy carries no PATH export, so if a foreign vastai now resolves
+    # first the block must be rewritten to reassert precedence, not skipped.
+    rewrite=""
     if [ -n "$rc_file" ] && [ -f "$rc_file" ] && grep -qF "$marker" "$rc_file"; then
-        return 0  # already configured; takes effect in new shells
+        if [ -z "$path_line" ] || grep -qF "$path_export" "$rc_file"; then
+            return 0  # already configured; takes effect in new shells
+        fi
+        grep -qF "$marker_end" "$rc_file" || return 0  # block corrupt — coexistence warning covers it
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Declined as not worth the branching: (1) the markers are shell comments, so a missing end marker breaks nothing at runtime — the export/completion lines in a "partial" block still execute; (2) the early return on those cases is byte-identical to pre-PR behavior — this PR only added the end-marker guard on the new rewrite path; (3) in the one case that matters (precedence needed, block not safely rewritable) `RC_UPDATED` stays unset, so `check_pip_coexistence` already fires the warning. The state is only reachable by hand-deleting the end-marker line — the installer writes the block in a single append.

File: scripts/install.sh [hunk 3]
```diff
@@ -270,32 +271,57 @@ setup_path() {
...
+        rewrite=1
     fi
 
     # Default-on, no prompt — but never edit rc under --no-modify-path or
     # without a TTY (CI/pipes); there we just print the block.
     if [ -z "$NO_MODIFY_PATH" ] && [ -n "$rc_file" ] && is_interactive; then
+        if [ -n "$rewrite" ]; then
+            awk -v s="$marker" -v e="$marker_end" '$0==s{skip=1;next} $0==e{skip=0;next} !skip' \
+                "$rc_file" > "$rc_file.vastai.tmp" && mv "$rc_file.vastai.tmp" "$rc_file"
+        fi
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Fixed in 82eb6a0 — good catch. The rewrite now writes through the rc file with `cat tmp > rc_file` (preserving symlink/inode/perms for dotfile-manager setups), and on awk/cat failure it cleans up the tmp file and falls back to printing the block for manual setup instead of appending a duplicate (`RC_UPDATED` stays unset, so the coexistence warning also fires). `test_pip_shadowing_rerun` now runs with `.bashrc` as a symlink and asserts the link survives, the target gets the rewrite, and no tmp file is left behind.

File: scripts/publish_release.py [hunk 4]
```diff
@@ -177,8 +88,11 @@ def attach_release(tag, manifest_dir, wheel, *, dry):
             f"--clobber {Path(wheel).name} manifest.json manifest.env install.sh; "
             f"gh release edit {tag} --draft=false")
         return
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/publish_release.py]: Fixed in 82eb6a0 by deleting the hand-written summary rather than expanding it: dry mode now skips the `gh` probe (assumes create — no gh/auth needed to dry-run) and the three real command lists flow through `run(..., dry=dry)`, which prints exact argv. Dry-run output can no longer drift from the real invocation:

```
[dry-run] gh release create v1.2.3 --draft --title v1.2.3 --notes vastai 1.2.3 --generate-notes
[dry-run] gh release upload v1.2.3 --clobber dist/vastai-1.2.3-<built>.whl /tmp/.../manifest.json /tmp/.../manifest.env .../scripts/install.sh
[dry-run] gh release edit v1.2.3 --draft=false
```

**Review by vastzuby — APPROVED** [review]

File: scripts/install.sh [hunk 5]
```diff
@@ -270,32 +271,66 @@ setup_path() {
...
 
+    # Marker idempotency is content-aware: an existing block written when PATH
+    # was healthy carries no PATH export, so if a foreign vastai now resolves
+    # first the block must be rewritten to reassert precedence, not skipped.
+    rewrite=""
     if [ -n "$rc_file" ] && [ -f "$rc_file" ] && grep -qF "$marker" "$rc_file"; then
-        return 0  # already configured; takes effect in new shells
+        if [ -z "$path_line" ] || grep -qF "$path_export" "$rc_file"; then
+            return 0  # already configured; takes effect in new shells
+        fi
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Fixed in 10c4abb — confirmed real: a re-run in a still-shadowed shell warned about a coexistence problem the rc had already resolved. The early-return path now sets `RC_UPDATED=1` (with a comment pinning its meaning as "the rc guarantees precedence", not "written this run"), which suppresses the spurious warning and lets the "Use it now" hint print for the still-shadowed current shell. `test_pip_shadowing_rerun` gained a third install run asserting the re-run is a quiet no-op: rc byte-identical, no warning, hint still shown.

File: scripts/install.sh [hunk 6]
```diff
@@ -270,32 +271,71 @@ setup_path() {
...
     fi
 
     # Default-on, no prompt — but never edit rc under --no-modify-path or
     # without a TTY (CI/pipes); there we just print the block.
     if [ -z "$NO_MODIFY_PATH" ] && [ -n "$rc_file" ] && is_interactive; then
+        if [ -n "$rewrite" ]; then
+            # cat, not mv: rc files are often symlinks (dotfile managers), and
+            # writing through the link preserves its inode and permissions.
+            if awk -v s="$marker" -v e="$marker_end" '$0==s{skip=1;next} $0==e{skip=0;next} !skip' \
+                "$rc_file" > "$rc_file.vastai.tmp" && cat "$rc_file.vastai.tmp" > "$rc_file"; then
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Fixed in a985b14, inverted: the installer always writes markers at column 0 (the block string isn't indented), so an indented end marker only arises from hand-editing or CRLF conversion. Rather than make awk tolerant of malformed blocks, the guard is now strict — both markers must match as exact lines (`grep -qxF`) or the rewrite is refused and falls back to printed instructions + warning. New `test_rc_block_malformed` pins the fail-safe (rc byte-identical, warning printed).

File: scripts/install.sh [hunk 7]
```diff
@@ -270,32 +271,71 @@ setup_path() {
...
+    # line anyway so this install takes precedence in new shells.
+    local existing
+    existing="$(command -v vastai 2>/dev/null || true)"
+    case "$existing" in
+        ""|"$LOCAL_BIN/vastai"|"$ROOT/bin/vastai") ;;
+        *) PATH_OK="" ;;
+    esac
 
     # rc block: completion always (sources the static script); PATH export only
     # if $LOCAL_BIN isn't already on PATH — so we never skip completion on PATH_OK.
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Fixed in a985b14 — comment now reads: PATH export whenever $LOCAL_BIN precedence isn't guaranteed (off PATH or outranked).

**Comment by sammy-vastai** [comment]
Reworked the rc handling in e795505 to the rustup/uv env-file pattern: one constant rc line sourcing `~/.vastai/env.sh`, append-only, never rewritten. The marker-block rewrite machinery from earlier rounds is deleted — precedence is now asserted at every shell startup by env.sh, which also makes later pip installs self-resolving. Net −68 lines; PR body updated.

File: scripts/publish_release.py [hunk 8]
```diff
@@ -72,93 +61,15 @@ def run(cmd, *, dry, capture=False, check=True, env=None):
...
-
-
-def wait_for_pypi(version, dry):
-    """Poll PyPI until the version is downloadable."""
-    url = f"https://pypi.org/pypi/{PACKAGE}/{version}/json"
+def resolve_wheel(version, *, dry):
+    """The wheel poetry just built/published, in dist/ — hash those exact bytes."""
+    matches = sorted(glob.glob(str(REPO_ROOT / "dist" / f"{PACKAGE}-{version}-*.whl")))
+    if matches:
+        return matches[0]
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/publish_release.py:71]: Fixed in a1766f4 — >1 match now raises with the candidate names, matching make_manifest.py's exactly-one contract. (Unreachable in CI's fresh checkout, but a stale local dist/ is real.)

File: scripts/install.sh [hunk 9]
```diff
@@ -262,45 +263,46 @@ generate_completions() {
...
 
-    case ":$PATH:" in
-        *":$LOCAL_BIN:"*) PATH_OK=1 ;;
-        *) PATH_OK="" ;;
+    # Hint "Use it now" whenever this shell wouldn't run our vastai (off PATH or outranked, e.g. by a pip install).
+    local existing
+    existing="$(command -v vastai 2>/dev/null || true)"
+    case "$existing" in
+        "$LOCAL_BIN/vastai"|"$ROOT/bin/vastai") ;;
+        *) PATH_PREPENDED=1 ;;
```
[the remark this answers is not in the document]
Comment (sammy-vastai) [reply scripts/install.sh]: Fixed in a1766f4 — renamed to NEEDS_PATH_HINT; agreed the old name stopped being true once env.sh took over the prepending.
